From: Bob Moore <robert.moore@intel.com>
Date: Thu, 10 Apr 2008 15:06:43 +0000 (+0400)
Subject: ACPICA: Fix for possible error when packages/buffers are passed to methods externally
X-Git-Tag: v2.6.26-rc1~101^2~12
X-Git-Url: http://git.kernel.org/?p=linux%2Fkernel%2Fgit%2Ftorvalds%2Flinux-2.6.git;a=commitdiff_plain;h=24a3157a90ddf851a0880c0b8963bc43481cd85b

ACPICA: Fix for possible error when packages/buffers are passed to methods externally

Fixed a problem where buffer and package objects passed as
arguments to a control method via the external AcpiEvaluateObject
interface could cause an AE_AML_INTERNAL exception depending on the
order and type of operators executed by the target control method.

Signed-off-by: Bob Moore <robert.moore@intel.com>
Signed-off-by: Alexey Starikovskiy <astarikovskiy@suse.de>
Signed-off-by: Len Brown <len.brown@intel.com>
---

Index: 2008-devel/drivers/acpi/utilities/utcopy.c
===================================================================
--- 2008-devel.orig/drivers/acpi/utilities/utcopy.c
+++ 2008-devel/drivers/acpi/utilities/utcopy.c
@@ -502,6 +502,10 @@ acpi_ut_copy_esimple_to_isimple(union ac
 			    external_object->buffer.length);
 
 		internal_object->buffer.length = external_object->buffer.length;
+
+		/* Mark buffer data valid */
+
+		internal_object->buffer.flags |= AOPOBJ_DATA_VALID;
 		break;
 
 	case ACPI_TYPE_INTEGER:
@@ -577,6 +581,10 @@ acpi_ut_copy_epackage_to_ipackage(union 
 		}
 	}
 
+	/* Mark package data valid */
+
+	package_object->package.flags |= AOPOBJ_DATA_VALID;
+
 	*internal_object = package_object;
 	return_ACPI_STATUS(status);
 }
